cpp/src/quantiles/groupby.hpp

```diff
@@ -0,0 +1,159 @@
+
+  groupby(cudf::table const& keys, bool include_nulls = false,
+          cudaStream_t stream = 0)
+  : _keys(keys)
+  , _num_keys(keys.num_rows())
+  , _include_nulls(include_nulls)
+  , _stream(stream)
+  {};
+
+  ~groupby() {}
+
+  /**
+   * @brief Group and sort a column of values
+   * 
+   * Sorts and groups the @p values where the groups are dictated by key table
+   * and the elements are sorted ascending within the groups. Calculates and
+   * returns the number of valid values within each group. 
+   * 
+   * @note Size of @p values should be equal to number of rows in keys
+   * 
+   * @param values The value column to group and sort
+   * @return the sorted and grouped column and per-group valid count
+   */
+  std::pair<gdf_column, rmm::device_vector<gdf_size_type> >
+  sort_values(gdf_column const& values);
+
+  /**
+   * @brief Get a table of sorted unique keys
+   * 
+   * @return a new table in which each row is a unique row in the sorted key table.
+   */
+  cudf::table unique_keys();
+
+  /**
+   * @brief Get the number of groups in `keys`
+   * 
+   */
+  gdf_size_type num_groups() { return group_offsets().size(); }
+
+  /**
+   * @brief Return the effective number of keys
+   * 
+   * When include_nulls = true, returned value is same as `keys.num_rows()`
+   * When include_nulls = false, returned value is the number of rows in `keys`
+   *  in which no element is null
+   */
+  gdf_size_type num_keys() { return _num_keys; }
+
+  /**
+   * @brief Get the sorted order of `keys`.
+   *
+   * Gathering `keys` by sort order indices will produce the sorted key table.
+   * 
+   * Computes and stores the key sorted order on first invocation, and returns the
+   * stored order on subsequent calls.
+   * 
+   * @return the sort order indices for `keys`.
+   */
+  gdf_column const& key_sort_order();
+
+  /**
+   * @brief Get the group offsets.
+   * 
+   * Computes and stores the group offsets on first invocation and returns
+   * the stored group offsets on subsequent calls.
+   * 
+   * @return vector of offsets of the starting point of each group in the sorted key table
+   */
+  index_vector const& group_offsets();
+
+  /**
+   * @brief Get the group labels
+   * 
+   * For a row in `keys`, its group label is the group which it belongs to
+   * 
+   * Computes and stores labels on first invocation and returns stored labels on
+   * subsequent calls.
+   * 
+   * @return vector of group labels for each row in the sorted key column
+   */
+  index_vector const& group_labels();
+
+  /**
+   * @brief Get the unsorted labels
+   * 
+   * For each row in `keys`, the unsorted label is the group it would
+   * belong to after sorting.
+   * 
+   * Computes and stores unsorted labels on first invocation and returns stored
+   * labels on subsequent calls.
+   * 
+   * @return column of group labels in the order of the unsorted key table
+   */
+  gdf_column const& unsorted_labels();
```

Comment:
It seems odd for `unsorted_labels` to return a `gdf_column` when `group_labels` returns an `index_vector`. Can they be made consistent?

Reply:
`unsorted_labels` is mainly intended for internal use to sort columns so it returns a `gdf_column` to avoid converting it to a column everytime we need to combine it and a values column to run `gdf_order_by`. That would result in a messy code which i already am tired of seeing eveywhere. Hopefully, not after the column refactor. `group_labels` is for use by aggregations that receive a sorted value column. And I foresee most of them to use thrust `_by_key` functions for it so again it'd result in a messy conversion into a device vector. The reason `unsorted_labels()` is not private is because I'm not trying to police all aggs to use what I'll let them use.